File: tools/operator/observer-debug.js
#!/usr/bin/env node
/**
 * Observer Debugging Utilities — OCE-3.17
 * 
 * CLI tools for inspecting and debugging OCE Observer Runtime.
 * 
 * Usage:
 *   node tools/operator/observer-debug.js list              # List all observers
 *   node tools/operator/observer-debug.js status <id>       # Observer status
 *   node tools/operator/observer-debug.js health <id>       # Health metrics
 *   node tools/operator/observer-debug.js events <id>       # Recent events for observer
 *   node tools/operator/observer-debug.js logs <id>         # Observer logs
 *   node tools/operator/observer-debug.js all               # Full system overview
 */

const http = require('http');

const OCE_HOST = process.env.OCE_HOST || '127.0.0.1';
const OCE_PORT = process.env.OCE_PORT || 8000;

function request(method, path, body = null) {
    return new Promise((resolve) => {
        const bodyData = body ? JSON.stringify(body) : null;
        const options = {
            hostname: OCE_HOST, port: OCE_PORT,
            path, method,
            headers: { 'Content-Type': 'application/json' },
            timeout: 10000,
        };
        if (bodyData) options.headers['Content-Length'] = Buffer.byteLength(bodyData);
        const req = http.request(options, (res) => {
            let data = '';
            res.on('data', c => data += c);
            res.on('end', () => {
                try { resolve({ ok: res.statusCode < 400, status: res.statusCode, data: JSON.parse(data) }); }
                catch (e) { resolve({ ok: res.statusCode < 400, status: res.statusCode, data }); }
            });
        });
        req.on('error', e => resolve({ ok: false, error: e.message }));
        req.on('timeout', () => { req.destroy(); resolve({ ok: false, error: 'timeout' }); });
        if (bodyData) req.write(bodyData);
        req.end();
    });
}

const C = { r: '\x1b[0m', red: '\x1b[31m', grn: '\x1b[32m', yel: '\x1b[33m', blu: '\x1b[34m', mag: '\x1b[35m', cyn: '\x1b[36m', gry: '\x1b[90m', bld: '\x1b[1m' };
const sColor = s => s === 'active' ? C.grn : s === 'suspended' ? C.yel : s === 'destroyed' ? C.red : C.gry;
const hColor = h => h ? C.grn : C.red;

async function cmd_list() {
    console.log(`${C.cyn}👁 Observer List${C.r}\n`);
    const r = await request('GET', '/observers');
    if (!r.ok) { console.log(`${C.red}❌ ${r.error || r.status}${C.r}`); return; }
    const observers = r.data || [];
    if (!observers.length) { console.log(`${C.gry}No observers found.${C.r}`); return; }
    console.log(`${C.bld}${'ID'.padEnd(20)} ${'Type'.padEnd(15)} ${'State'.padEnd(12)} ${'Health'.padEnd(10)} Entropy${C.r}`);
    console.log('─'.repeat(70));
    for (const o of observers) {
        const sc = sColor(o.state);
        const hc = hColor(o.healthy);
        console.log(`${(o.observer_id || o.id || '?').padEnd(20)} ${(o.type || '?').padEnd(15)} ${sc}${(o.state || '?').padEnd(12)}${C.r} ${hc}${(o.healthy ? '✅' : '❌').padEnd(10)}${C.r} ${o.entropy?.toFixed(2) || 'N/A'}`);
    }
    console.log(`\n  Total: ${observers.length} observers`);
}

async function cmd_status(id) {
    if (!id) { console.log(`${C.red}Usage: observer status <id>${C.r}`); return; }
    console.log(`${C.cyn}👁 Observer Status: ${id}${C.r}\n`);
    const r = await request('GET', `/observers/${id}`);
    if (!r.ok) { console.log(`${C.red}❌ ${r.error || r.status}${C.r}`); return; }
    const o = r.data || {};
    const sc = sColor(o.state);
    const hc = hColor(o.healthy);
    console.log(`  ID:       ${o.observer_id || o.id || id}`);
    console.log(`  Type:     ${o.type || 'N/A'}`);
    console.log(`  State:    ${sc}${o.state || 'N/A'}${C.r}`);
    console.log(`  Healthy:  ${hc}${o.healthy ? 'Yes' : 'No'}${C.r}`);
    console.log(`  Entropy:  ${o.entropy?.toFixed(4) || 'N/A'}`);
    console.log(`  Task:     ${o.task || 'N/A'}`);
    if (o.config) console.log(`  Config:   ${JSON.stringify(o.config, null, 2)}`);
    if (o.created_at) console.log(`  Created:  ${o.created_at}`);
    if (o.last_active) console.log(`  Active:   ${o.last_active}`);
}

async function cmd_health(id) {
    if (!id) { console.log(`${C.red}Usage: observer health <id>${C.r}`); return; }
    console.log(`${C.cyn}🏥 Observer Health: ${id}${C.r}\n`);
    const r = await request('GET', `/observers/${id}/health`);
    if (!r.ok) { console.log(`${C.red}❌ ${r.error || r.status}${C.r}`); return; }
    const h = r.data || {};
    const hc = hColor(h.healthy);
    console.log(`  Healthy:      ${hc}${h.healthy ? 'Yes' : 'No'}${C.r}`);
    console.log(`  Entropy:      ${h.entropy?.toFixed(4) || 'N/A'}`);
    console.log(`  Drift:        ${h.drift?.toFixed(4) || 'N/A'}`);
    console.log(`  Budget:       ${h.budget_remaining?.toFixed(1) || 'N/A'}`);
    console.log(`  Last check:   ${h.last_check || 'N/A'}`);
    if (h.issues && h.issues.length > 0) {
        console.log(`\n  ${C.yel}Issues:${C.r}`);
        for (const issue of h.issues) console.log(`    ⚠ ${issue}`);
    }
}

async function cmd_events(id) {
    if (!id) { console.log(`${C.red}Usage: observer events <id>${C.r}`); return; }
    console.log(`${C.cyn}📡 Events for: ${id}${C.r}\n`);
    const r = await request('GET', `/events?source=${id}&limit=20`);
    if (!r.ok) { console.log(`${C.red}❌ ${r.error || r.status}${C.r}`); return; }
    const events = r.data || [];
    if (!events.length) { console.log(`${C.gry}No events found for observer: ${id}${C.r}`); return; }
    for (const e of events) {
        const pc = e.priority >= 3 ? C.red : e.priority === 2 ? C.yel : e.priority === 1 ? C.cyn : C.gry;
        const t = new Date(e.timestamp).toLocaleTimeString();
        console.log(`  ${C.gry}[${t}]${C.r} ${pc}[${e.priority}]${C.r} ${C.yel}${e.event_type}${C.r} ${JSON.stringify(e.payload).substring(0, 100)}`);
    }
    console.log(`\n  Total: ${events.length} events`);
}

async function cmd_logs(id) {
    if (!id) { console.log(`${C.red}Usage: observer logs <id>${C.r}`); return; }
    console.log(`${C.cyn}📋 Logs for: ${id}${C.r}\n`);
    const r = await request('GET', `/observers/${id}/logs`);
    if (!r.ok) { console.log(`${C.red}❌ ${r.error || r.status}${C.r}`); return; }
    const logs = r.data || [];
    if (!logs.length) { console.log(`${C.gry}No logs found.${C.r}`); return; }
    for (const log of logs) {
        const lvl = log.level === 'error' ? C.red : log.level === 'warn' ? C.yel : C.gry;
        console.log(`  ${lvl}[${log.level?.toUpperCase() || 'INFO'}]${C.r} ${log.message || log}`);
    }
}

async function cmd_all() {
    console.log(`${C.cyn}👁 OCE Observer Runtime — Full Overview${C.r}\n`);

    // Backend health
    const bh = await request('GET', '/health');
    console.log(`  Backend:     ${bh.ok ? C.grn + '✅ healthy' : C.red + '❌ ' + bh.error}${C.r}`);

    // SRRA-OPH
    const sh = await request('GET', '/health/srrs');
    console.log(`  SRRA-OPH:    ${sh.ok ? C.grn + '✅ healthy' : C.red + '❌ ' + sh.error}${C.r}`);

    // Event Fabric
    const eh = await request('GET', '/events/stats');
    console.log(`  Event Fabric: ${eh.ok ? C.grn + '✅' : C.red + '❌'} ingested=${eh.data?.total_ingested || 0}, history=${eh.data?.history_size || 0}${C.r}`);

    // Observers
    const r = await request('GET', '/observers');
    if (r.ok) {
        const observers = r.data || [];
        console.log(`\n  ${C.bld}Observers (${observers.length}):${C.r}`);
        if (!observers.length) console.log(`    ${C.gry}No observers created yet.${C.r}`);
        for (const o of observers) {
            const sc = sColor(o.state);
            const hc = hColor(o.healthy);
            console.log(`    ${(o.observer_id || o.id || '?').padEnd(20)} ${sc}${(o.state || '?').padEnd(12)}${C.r} ${hc}${o.healthy ? '✅' : '❌'}${C.r}`);
        }
    } else {
        console.log(`\n  Observers: ${C.yel}API not available yet (CC building OCE-3.1)${C.r}`);
    }
}

// ── CLI ──────────────────────────────────────────────────────────────────────

const [,, cmd, ...args] = process.argv;

(async () => {
    switch (cmd) {
        case 'list': await cmd_list(); break;
        case 'status': await cmd_status(args[0]); break;
        case 'health': await cmd_health(args[0]); break;
        case 'events': await cmd_events(args[0]); break;
        case 'logs': await cmd_logs(args[0]); break;
        case 'all': await cmd_all(); break;
        default:
            console.log(`${C.cyn}Observer Debug Utilities${C.r}\n`);
            console.log(`  ${C.bld}Commands:${C.r}`);
            console.log(`  ${C.grn}list${C.r}                         List all observers`);
            console.log(`  ${C.grn}status${C.r}  <id>                 Observer status`);
            console.log(`  ${C.grn}health${C.r}  <id>                 Health metrics`);
            console.log(`  ${C.grn}events${C.r}  <id>                 Recent events`);
            console.log(`  ${C.grn}logs${C.r}    <id>                 Observer logs`);
            console.log(`  ${C.grn}all${C.r}                          Full system overview`);
    }
})();

File: tools/operator/observer-debug.py
#!/usr/bin/env python3
"""OCE Observer Debug CLI — inspect and manage observers via the OCE backend API."""

import argparse
import sys
import urllib.request
import urllib.error
import json

BASE_URL = "http://localhost:8000"

# ── colour helpers ──────────────────────────────────────────────────────────

def _c(code: str, text: str) -> str:
    return f"\033[{code}m{text}\033[0m"

def green(text: str)  -> str: return _c("32", text)
def yellow(text: str) -> str: return _c("33", text)
def red(text: str)    -> str: return _c("31", text)
def bold(text: str)   -> str: return _c("1", text)
def dim(text: str)    -> str: return _c("2", text)

def health_color(state: str) -> str:
    s = state.lower()
    if s in ("healthy", "active", "ok"):
        return green(state)
    if s in ("warning", "degraded", "suspended"):
        return yellow(state)
    return red(state)

# ── HTTP helper ─────────────────────────────────────────────────────────────

def api_get(path: str) -> dict:
    url = f"{BASE_URL}{path}"
    try:
        req = urllib.request.Request(url)
        with urllib.request.urlopen(req, timeout=5) as resp:
            body = resp.read().decode()
            return json.loads(body) if body else {}
    except urllib.error.URLError as exc:
        print(red(f"✗  API unreachable at {url}"))
        print(dim(f"   {exc}"))
        sys.exit(1)
    except json.JSONDecodeError:
        print(red(f"✗  Invalid JSON response from {url}"))
        sys.exit(1)

# ── tabulate fallback (no external dep) ─────────────────────────────────────

def render_table(headers: list[str], rows: list[list[str]]) -> str:
    """Simple table renderer — uses tabulate if available, else manual."""
    try:
        from tabulate import tabulate as _tb
        return _tb(rows, headers=headers, tablefmt="simple")
    except ImportError:
        widths = [len(h) for h in headers]
        for row in rows:
            for i, cell in enumerate(row):
                widths[i] = max(widths[i], len(str(cell)))
        lines: list[str] = []
        header_line = "  ".join(h.ljust(widths[i]) for i, h in enumerate(headers))
        lines.append(bold(header_line))
        lines.append(dim("  ".join("-" * w for w in widths)))
        for row in rows:
            lines.append("  ".join(str(c).ljust(widths[i]) for i, c in enumerate(row)))
        return "\n".join(lines)

# ── command implementations ─────────────────────────────────────────────────

def cmd_list(_args: argparse.Namespace) -> None:
    data = api_get("/api/v1/observers")
    observers = data if isinstance(data, list) else data.get("observers", data.get("data", []))
    if not observers:
        print(yellow("No observers found."))
        return
    rows = []
    for o in observers:
        state = health_color(o.get("state", o.get("status", "unknown")))
        rows.append([
            str(o.get("id", "")),
            str(o.get("type", "")),
            state,
            str(o.get("health", o.get("health_score", "-"))),
        ])
    print(render_table(["ID", "Type", "State", "Health"], rows))
    print(dim(f"\n{len(observers)} observer(s)"))

def cmd_status(args: argparse.Namespace) -> None:
    data = api_get(f"/api/v1/observers/{args.id}")
    if not data:
        print(red(f"Observer '{args.id}' not found."))
        sys.exit(1)
    print(bold(f"Observer: {args.id}"))
    print(dim("─" * 40))
    for key, val in data.items():
        if isinstance(val, dict):
            print(f"  {bold(key)}:")
            for k2, v2 in val.items():
                print(f"    {k2}: {v2}")
        elif isinstance(val, list):
            print(f"  {bold(key)}: [{len(val)} items]")
        else:
            print(f"  {bold(key)}: {val}")

def cmd_health(args: argparse.Namespace) -> None:
    data = api_get(f"/api/v1/observers/{args.id}/health")
    if not data:
        print(red(f"Observer '{args.id}' not found."))
        sys.exit(1)
    entropy  = data.get("entropy", "-")
    drift    = data.get("drift", "-")
    budget   = data.get("budget", data.get("budget_remaining", "-"))
    score    = data.get("health_score", data.get("score", "-"))

    def _metric_color(val, low, high):
        try:
            v = float(val)
            if v <= low:  return green(str(val))
            if v <= high: return yellow(str(val))
            return red(str(val))
        except (TypeError, ValueError):
            return str(val)

    rows = [
        ["Health Score", _metric_color(score, 70, 40)],
        ["Entropy",     _metric_color(entropy, 0.3, 0.6)],
        ["Drift",       _metric_color(drift, 0.2, 0.5)],
        ["Budget",      str(budget)],
    ]
    print(bold(f"Health — {args.id}"))
    print(render_table(["Metric", "Value"], rows))

def cmd_events(args: argparse.Namespace) -> None:
    limit = getattr(args, "limit", 20)
    data = api_get(f"/api/v1/observers/{args.id}/events?limit={limit}")
    events = data if isinstance(data, list) else data.get("events", data.get("data", []))
    if not events:
        print(yellow(f"No events for observer '{args.id}'."))
        return
    rows = []
    for e in events:
        ts   = str(e.get("timestamp", e.get("ts", "")))
        etype = str(e.get("type", e.get("event_type", "")))
        detail = str(e.get("detail", e.get("message", "")))[:60]
        rows.append([ts, etype, detail])
    print(render_table(["Timestamp", "Type", "Detail"], rows))
    print(dim(f"\n{len(events)} event(s)"))

def cmd_logs(args: argparse.Namespace) -> None:
    lines = getattr(args, "lines", 50)
    data = api_get(f"/api/v1/observers/{args.id}/logs?lines={lines}")
    entries = data if isinstance(data, list) else data.get("logs", data.get("data", []))
    if not entries:
        print(yellow(f"No log entries for observer '{args.id}'."))
        return
    for entry in entries:
        if isinstance(entry, dict):
            ts      = entry.get("timestamp", entry.get("ts", ""))
            level   = str(entry.get("level", "")).upper()
            message = entry.get("message", entry.get("msg", ""))
            level_colored = {
                "INFO": green(level), "WARN": yellow(level),
                "ERROR": red(level), "DEBUG": dim(level),
            }.get(level, level)
            print(f"  {dim(ts)}  {level_colored}  {message}")
        else:
            print(f"  {entry}")
    print(dim(f"\n{len(entries)} log line(s)"))

def _post_command(action: str, id_: str) -> None:
    url = f"{BASE_URL}/api/v1/observers/{id_}/{action}"
    try:
        req = urllib.request.Request(url, method="POST",
                                     data=b"{}",
                                     headers={"Content-Type": "application/json"})
        with urllib.request.urlopen(req, timeout=5) as resp:
            body = resp.read().decode()
            data = json.loads(body) if body else {}
    except urllib.error.URLError as exc:
        print(red(f"✗  API error: {exc}"))
        sys.exit(1)
    state = data.get("state", data.get("status", action + "d"))
    print(green(f"✓  Observer '{id_}' {state}"))
    if data:
        for k, v in data.items():
            if k not in ("state", "status"):
                print(f"   {k}: {v}")

def cmd_activate(args: argparse.Namespace) -> None:
    _post_command("activate", args.id)

def cmd_suspend(args: argparse.Namespace) -> None:
    _post_command("suspend", args.id)

# ── CLI setup ───────────────────────────────────────────────────────────────

def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog="observer-debug",
        description="OCE Observer Debug CLI — inspect and manage observers",
    )
    sub = parser.add_subparsers(dest="command")

    sub.add_parser("list", help="List all observers")

    p = sub.add_parser("status", help="Full observer details")
    p.add_argument("id", help="Observer ID")

    p = sub.add_parser("health", help="Health metrics")
    p.add_argument("id", help="Observer ID")

    p = sub.add_parser("events", help="Recent events")
    p.add_argument("id", help="Observer ID")
    p.add_argument("limit", nargs="?", type=int, default=20, help="Max events (default 20)")

    p = sub.add_parser("logs", help="Log entries")
    p.add_argument("id", help="Observer ID")
    p.add_argument("lines", nargs="?", type=int, default=50, help="Max lines (default 50)")

    p = sub.add_parser("activate", help="Activate a suspended observer")
    p.add_argument("id", help="Observer ID")

    p = sub.add_parser("suspend", help="Suspend an active observer")
    p.add_argument("id", help="Observer ID")

    return parser

def main() -> None:
    parser = build_parser()
    args = parser.parse_args()

    if not args.command:
        parser.print_help()
        sys.exit(0)

    dispatch = {
        "list":     cmd_list,
        "status":   cmd_status,
        "health":   cmd_health,
        "events":   cmd_events,
        "logs":     cmd_logs,
        "activate": cmd_activate,
        "suspend":  cmd_suspend,
    }
    handler = dispatch.get(args.command)
    if handler:
        handler(args)
    else:
        parser.print_help()

if __name__ == "__main__":
    main()
